pkg/datapath/linux/config/config.go

```diff
@@ -98,6 +98,83 @@ func writeIncludes(w io.Writer) (int, error) {
 	return fmt.Fprintf(w, "#include \"lib/utils.h\"\n\n")
 }
 
+// WriteClusterConfig writes the configuration of cluster-wide options into the
+// specified writer.
+func (h *HeaderfileWriter) WriteClusterConfig(w io.Writer, cfg datapath.ClusterConfiguration) error {
+	cDefinesMap := cfg.GetOptions()
+	fw := bufio.NewWriter(w)
+
+	// Since golang maps are unordered, we sort the keys in the map
+	// to get a consistent written format to the writer. This maintains
+	// the consistency when we try to calculate hash for a datapath after
+	// writing the config.
+	keys := make([]string, 0, len(cDefinesMap))
+	for key := range cDefinesMap {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+
+	for _, key := range keys {
+		fmt.Fprintf(fw, "#define %s %s\n", key, cDefinesMap[key])
+	}
+
+	// Write the JSON encoded config as base64 encoded commented string to
+	// the header file.
+	jsonBytes, err := json.Marshal(cDefinesMap)
+	if err == nil {
+		// We don't care if some error occurs while marshaling the map.
+		// In such cases we skip embedding the base64 encoded JSON configuration
+		// to the writer.
+		encodedConfig := base64.StdEncoding.EncodeToString(jsonBytes)
+		fmt.Fprintf(fw, "\n// JSON_OUTPUT: %s\n", encodedConfig)
+	}
+
+	return fw.Flush()
+}
+
+// HeaderfileReader is a wrapper type which implements datapath.ConfigReader.
+// It manages reading and validation of datapath program headerfiles.
+type HeaderfileReader struct{}
+
+func NewHeaderfileReader() *HeaderfileReader {
+	return &HeaderfileReader{}
+}
+
+// ValdateClusterConfig reads an existing cluster-wide options config file and
+// validates it against the given configuration. This config file is immutable
+// and its values should not be changed.
+func (h *HeaderfileReader) ValidateClusterConfig(r io.Reader, cfg datapath.ClusterConfiguration) error {
+	cDefinesMap := cfg.GetOptions()
+
+	for key, currentVal := range h.ReadConfigMacros(r) {
+		if newVal, ok := cDefinesMap[key]; ok && newVal != currentVal {
+			return fmt.Errorf("cluster config file cannot be changed. %s was defined as %s, but new agent configuration would change it to %v", key, currentVal, newVal)
+		}
+	}
+
+	return nil
+}
+
+// ReadConfigMacros reads the given headerfile and returns a map of the macro definitions.
+func (h *HeaderfileReader) ReadConfigMacros(r io.Reader) map[string]string {
```

Comment:
As mentioned in https://github.com/cilium/cilium/pull/27520/files#r1346120795, I think this is problematic. We shouldn't treat headers as a state file, nor should we validate user configuration this deep into the code.

Reply:
thanks for taking another look, Timo

> Wouldn't it be simpler to remove the writable bit from cluster_config.h if we want to prevent future modifications?

Definitely simpler, but I thought this approach would be more flexible, allowing additional macros to be defined or repairing a header with a broken config

> Do we have validating webhooks for configmap updates?

Not currently, but there's an open issue for that [here](https://github.com/cilium/cilium/issues/18652)

> Could you outline the exact issues you ran into adding these values to node_config.h?

I don't recall the exact issues offhand but I think it was related to defining macros in common.h that reference values from node_config.h. node_config.h is only sometimes included indirectly in common.h. I'm sure I could revisit and make that work if using the cluster_config header for validation is out.

It seems like we may not have a precedent for validating a running config against a previous one, and without that we may want to consider shelving this until we implement the [migration path](https://github.com/isovalent/roadmap/issues/830) for existing clusters (this was scoped into a separate roadmap item).
cc: @aanm @marseel for thoughts on this

Reply:
Instead of reading the header back out, is there an other oracle you could use to determine if a cluster was deployed with a smaller cluster limit? Cilium already leaves quite a bit of state on the node in the form of maps and files in `'/var/run/cilium/state`. There's also `agent-runtime-config.json` in there, for example.

Reply:
`agent-runtime-config.json` might be a good alternative! seems like I should spend some time exploring what secrets that directory contains :grin:

Reply:
@ti-mo dropped the validation and cluster_config.h. The new macros are now defined in node_config.h